packages/common/src/utils/sdkMigrationUtils.ts

```diff
@@ -0,0 +1,162 @@
+import { detailedDiff } from 'deep-object-diff'
+import { isEmpty } from 'lodash'
+
+import { ErrorLevel, ReportToSentryArgs } from '~/models/ErrorReporting'
+import { FeatureFlags, RemoteConfigInstance } from '~/services/remote-config'
+
+export type CheckSDKMigrationArgs<T> = {
+  legacy: T
+  migrated?: T | Error
+}
+
+export class SDKMigrationFailedError extends Error {
+  public endpointName: string
+  public innerMessage: string
+  public legacyValue: unknown
+  public migratedValue: unknown
+
+  constructor({
+    endpointName,
+    innerMessage,
+    legacyValue,
+    migratedValue
+  }: {
+    endpointName: string
+    innerMessage: string
+    legacyValue?: unknown
+    migratedValue?: unknown
+  }) {
+    super(`Diff ${endpointName} failed: ${innerMessage}`)
+    this.name = 'SDKMigrationFailedError'
+    this.endpointName = endpointName
+    this.innerMessage = innerMessage
+    this.legacyValue = legacyValue
+    this.migratedValue = migratedValue
+  }
+}
+
+/** Compares a legacy and migrated response, which must be the same shape. For
+ * literal values, will do a strict equals. For objects, will do a deep diff.
+ * Throws `SDKMigrationFailedError` if there is a difference between the two responses.
+ */
+export const compareSDKResponse = <T extends object>(
+  { legacy, migrated }: CheckSDKMigrationArgs<T>,
+  endpointName: string
+) => {
+  // Migrated is an error, skip the diff
+  if (migrated instanceof Error) {
+    throw new SDKMigrationFailedError({
+      endpointName,
+      innerMessage: 'Migrated response was error',
+      legacyValue: legacy,
+      migratedValue: migrated
+    })
+  }
+  // Both object-like, perform deep diff
+  if (typeof legacy === 'object' && typeof migrated === 'object') {
+    const diffValue = detailedDiff(legacy, migrated)
```

Comment:
we can just use lodash here instead right? `isEqual` performs deep comparison and we aren't using the added/deleted/updated

Reply:
If we just want a "are they both equal" approach, yes very true.

There's an updated implementation of this, but the short answer is that those fields _are_ used in examining what is actually different between two responses vs. having to hunt through them property by property on every level to determine what the difference is. They just aren't used programmatically.

Reply:
yeah... if they aren't used programmatically though, I don't think the future reader would understand why. Don't love adding another package _just_ for this (600 bytes is small, fair enough). can we report it to sentry? or keep it in as a debug tool? or add some console.debugs?

just trying to be devil's advocate in a world where decreasing bundle size is a big goal

Reply:
I appreciate it! We do send the diff up to sentry now. That was a change I added after opening the PR